tests/gpflow/test_monitor.py

```diff
@@ -289,3 +289,9 @@ def tf_func(step):
 
     for step in tf.range(100):
         tf_func(step)
+
+
+def test_scipy_monitor(monitor, model):
+    opt = gpflow.optimizers.Scipy()
+
+    opt.minimize(model.training_loss, model.trainable_variables, step_callback=monitor)
```

Comment:
This test does a basic check whether the code _runs_ (doesn't blow up) but it doesn't, for example, check whether monitor actually got called. You can use mocks for this - we've not been using them across GPflow yet, though I think that'd make a lot of tests faster and easier to understand! There's an overview here: https://changhsinlee.com/pytest-mock/

would you be interested in having a go at figuring that out yourself, or would you prefer if I take over that bit?

Reply:
Oooh, good point! So one solution is that I can write a simple custom step_callback function that progresses a step counter. Then, I can query the step and check that the counter has increased... this will ensure that a monitor is actually being called... But, it won't check if everything works properly in this particular scenario...

I've had a quick look at this link, thanks! That's cool. For these tests, what would we mock?

Reply:
Have made this change to check that the monitor does indeed get called...